static/js/server_events.js

```diff
@@ -104,7 +104,14 @@ function get_events_success(events) {
         try {
             messages = echo.process_from_server(messages);
             _.each(messages, message_store.set_message_booleans);
-            message_events.insert_new_messages(messages);
+            var sent_by_this_client = false;
+            _.each(messages, function (msg) {
+                var msg_state = sent_messages.messages[msg.local_id];
+                if (msg_state) {
+                    sent_by_this_client = true;
+                }
+            });
+            message_events.insert_new_messages(messages, sent_by_this_client);
```

Comment:
How does this avoid triggering the notice twice for messages that were locally echoed?  We might need to inspect some state on the `msg_state` pboject.

Reply:
In line number 105, `messages = echo.process_from_server(messages);` just returns non locally echoed messages. If there are locally echoed messages than the `echo.process_from_server(messages);` return an empty array. I have tested it manually using the console.log.
So if it is locally echoed message than the array returned by `echo.process_from_server(messages);` is empty so _.each will not run.
I first tried to check msg_state.data.locally_echoed but I realised that the echo.process_from_server(messages) is giving empty array for local echo messages.

Reply:
Also I have a side question. Do the server always sends individual messages or in bulk in form of array of messages ?

Reply:
@timabbott  If the server sends mix of messages like message not sent by client and messages that were locally echoed together in a single event than I think I have to track some state in `msg_state` otherwise if server send one message at a time than it is not be needed, right?

Reply:
> Do the server always sends messages one at a time or in bulk in form of array of messages ?

Usually one at a time, but could be in bulk if the client has been offline for a while.